base/libgit2/callbacks.jl

```diff
@@ -42,20 +42,38 @@ end
 function user_abort()
     # Note: Potentially it could be better to just throw a Julia error.
     ccall((:giterr_set_str, :libgit2), Void,
-          (Cint, Cstring),
-          Cint(Error.Callback), "Aborting, user cancelled credential request.")
-
+          (Cint, Cstring), Cint(Error.Callback),
+          "Aborting, user cancelled credential request.")
     return Cint(Error.EUSER)
 end
 
+function prompt_limit()
+    ccall((:giterr_set_str, :libgit2), Void,
+          (Cint, Cstring), Cint(Error.Callback),
+          "Aborting, maximum number of prompts reached.")
+    return Cint(Error.EAUTH)
+end
+
+function exhausted_abort()
+    ccall((:giterr_set_str, :libgit2), Void,
+          (Cint, Cstring), Cint(Error.Callback),
+          "All authentication methods have failed.")
+    return Cint(Error.EAUTH)
+end
+
 function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload, username_ptr)
     creds = Base.get(p.credential)::SSHCredentials
+    modified = false
```

Comment:
Some of these `modified` changes seem like they could simplified (e.g., You could probably just have `modified = p.first_pass && isfilled(creds)` at the beginning).

Reply:
It may be better if I drop `modified` and just do a comparison between the initial credentials and the current credentials. I'll do some checks.

Reply:
My idea of doing the comparison won't work. If a user were to enter the same invalid credentials twice in a row they would get an "All authentication methods have failed" error. The name `modified` isn't quite the right term as this just verifying that a new authentication method has been attempted.

Reply:
I reworked this a little bit. I've ended up changing `modified` to `revised`. I think this term is better suited for the current behaviour. As for `revised = p.first_pass && isfilled(creds)` I've decided to keep the `if` block as it matches the way `revised` is used in the reset of the code.